File: src/2015/06/day.06.ts
import fs from "fs";

enum Actions {
  OFF,
  ON,
  TOGGLE,
}

const ACTION_MAP = {
  off: Actions.OFF,
  on: Actions.ON,
  toggle: Actions.TOGGLE,
};

type Instruction = {
  action: Actions;
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
};

function parseInput() {
  return fs
    .readFileSync("src/inputs/2015/06/day.06.input.txt")
    .toString()
    .split("\n")
    .filter((x) => x)
    .map(parseInstruction);
}

function parseInstruction(value: string): Instruction {
  const action =
    ACTION_MAP[value.match(/on|off|toggle/)![0] as keyof typeof ACTION_MAP];

  const [minX, minY, maxX, maxY] = value.match(/\d+/g)!.map(Number);

  return { action, minX, minY, maxX, maxY };
}

export function part1() {
  return getTotalBrightness(calculateBinaryBrightness);
}

export function part2() {
  return getTotalBrightness(calculateCumulativeBrightness);
}

function getTotalBrightness(
  calculateBrightness: (previousBrightness: number, action: Actions) => number
) {
  const instructions = parseInput();

  let totalBrightness = 0;

  for (let x = 0; x < 1_000; x++) {
    for (let y = 0; y < 1_000; y++) {
      let brightness = 0;

      for (const instruction of instructions) {
        if (isApplicable(x, y, instruction)) {
          brightness = calculateBrightness(brightness, instruction.action);
        }
      }

      totalBrightness += brightness;
    }
  }

  return totalBrightness;
}

function calculateBinaryBrightness(
  previousBrightness: number,
  action: Actions
) {
  let currentBrightness: number;

  switch (action) {
    case Actions.ON: {
      currentBrightness = 1;
      break;
    }

    case Actions.OFF: {
      currentBrightness = 0;
      break;
    }

    case Actions.TOGGLE: {
      currentBrightness = 1 - previousBrightness;
      break;
    }
  }

  return currentBrightness;
}

function calculateCumulativeBrightness(
  previousBrightness: number,
  action: Actions
) {
  let currentBrightness: number;

  switch (action) {
    case Actions.ON: {
      currentBrightness = previousBrightness + 1;
      break;
    }

    case Actions.OFF: {
      currentBrightness = Math.max(0, previousBrightness - 1);
      break;
    }

    case Actions.TOGGLE: {
      currentBrightness = previousBrightness + 2;
      break;
    }
  }

  return currentBrightness;
}

function isApplicable(x: number, y: number, instruction: Instruction) {
  return (
    instruction.minX <= x &&
    x <= instruction.maxX &&
    instruction.minY <= y &&
    y <= instruction.maxY
  );
}

File: src/2015/06/day.06.py
from dataclasses import dataclass
from re import match
from timeit import default_timer as timer
from typing import Tuple


@dataclass
class Instruction:
    action: str
    top_left: Tuple[int, int]
    bottom_right: Tuple[int, int]


def to_instruction(groups: Tuple[str]):
    return Instruction(groups[0], (int(groups[1]), int(groups[2])), (int(groups[3]), int(groups[4])))


def parse_input():
    with open("src/inputs/2015/06/day.06.input.txt", "r") as file:
        return [to_instruction(match(r"(?:turn )?(on|off|toggle) (\d+),(\d+) through (\d+),(\d+)", line).groups()) for line in file.readlines()]


def part1():
    instructions = parse_input()

    count = 0

    for x in range(0, 1000):
        for y in range(0, 1000):
            state = 0

            for instruction in instructions:
                if in_rectange((x, y), instruction.top_left, instruction.bottom_right):
                    match instruction.action:
                        case "on":
                            state = 1
                        case "off":
                            state = 0
                        case "toggle":
                            state = 1 - state

            count += state

    return count


def part2():
    instructions = parse_input()

    total_brightness = 0

    for x in range(0, 1000):
        for y in range(0, 1000):
            brightness = 0

            for instruction in instructions:
                if in_rectange((x, y), instruction.top_left, instruction.bottom_right):
                    match instruction.action:
                        case "on":
                            brightness += 1
                        case "off":
                            brightness = max(0, brightness - 1)
                        case "toggle":
                            brightness += 2

            total_brightness += brightness

    return total_brightness


def in_rectange(point: Tuple[int, int], top_left: Tuple[int, int], bottom_right: Tuple[int, int]) -> bool:
    return point[0] >= top_left[0] and point[0] <= bottom_right[0] and point[1] >= top_left[1] and point[1] <= bottom_right[1]


def timed(fn):
    start = timer()
    result = fn()

    return (timer() - start, result)


(p1_elapsed, p1_answer) = timed(part1)
print(f"Part 1: {p1_answer} ({p1_elapsed:.3f}s)")

(p2_elapsed, p2_answer) = timed(part2)
print(f"Part 2: {p2_answer} ({p2_elapsed:.3f}s)")
